File: ruby/lib/dubs/pattern.rb
# frozen_string_literal: true

module Dubs
  module Pattern
    module_function

    # Interpolate a pattern template with values.
    #
    #   Pattern.interpolate("{adjective}-{noun}-{token}", adjective: "vicious", noun: "sazabi", token: "a3f1", separator: "-")
    #   # => "vicious-sazabi-a3f1"
    #
    def interpolate(template, adjective:, noun:, token:, separator: "-")
      result = template
        .gsub("{adjective}", adjective)
        .gsub("{noun}", noun)
        .gsub("{token}", token)

      # Replace literal hyphens in the template with the configured separator
      if separator != "-"
        result = result.gsub("-", separator)
      end

      # Clean up: remove trailing/leading separators if token is empty
      if token.empty?
        result = result
          .chomp(separator)
          .delete_prefix(separator)
          .gsub("#{separator}#{separator}", separator)
      end

      result
    end

    # Resolve a pattern name to its template string.
    def resolve(pattern)
      case pattern
      when Symbol, String
        Data.pattern(pattern)
      else
        pattern.to_s
      end
    end
  end
end

File: rust/src/pattern.rs
use crate::data::PATTERNS;

/// Named pattern presets.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PatternName {
    Default,
    Classified,
    Designation,
    Codename,
    Serial,
}

impl PatternName {
    pub fn template(&self) -> &str {
        let key = match self {
            PatternName::Default => "default",
            PatternName::Classified => "classified",
            PatternName::Designation => "designation",
            PatternName::Codename => "codename",
            PatternName::Serial => "serial",
        };
        PATTERNS.patterns.get(key).map(|s| s.as_str()).unwrap_or("{adjective}-{noun}-{token}")
    }
}

/// Interpolate a pattern template with values.
pub fn interpolate(template: &str, adjective: &str, noun: &str, token: &str, separator: &str) -> String {
    let mut result = template
        .replace("{adjective}", adjective)
        .replace("{noun}", noun)
        .replace("{token}", token);

    // Replace literal hyphens with custom separator
    if separator != "-" {
        result = result.replace('-', separator);
    }

    // Clean up trailing/leading separators if token is empty
    if token.is_empty() {
        let double = format!("{separator}{separator}");
        result = result
            .trim_end_matches(separator)
            .trim_start_matches(separator)
            .replace(&double, separator);
    }

    result
}
